### PR TITLE
Fixes #2251

### LINKED ISSUE #2251: Not able to use SignedAssertionFilePath for ClientCredentials
### Microsoft.Identity.Web Library

Microsoft.Identity.Web

### Microsoft.Identity.Web version

2.9.0

### Web app

Sign-in users and call web APIs

### Web API

Protected web APIs call downstream web APIs

### Token cache serialization

In-memory caches

### Description

A Web Api that is running in AKS is configured to use SignedAssertionFilePath for ClientCredentials. When the Web Api is calling down stream api, it is throwing the attached error. According to the [documentation ](https://github.com/AzureAD/microsoft-identity-web/wiki/v2.0) if the "SignedAssertionFileDiskPath" is not provided, it is picked from AZURE_FEDERATED_TOKEN_FILE environment variable. Can confirm that this variable, file it shows and the token in the file is present on the pod running the web app.

### Reproduction steps

1. Add below section in appsettings.json file of the WebApi
> "ClientCredentials": [
> {
> "SourceType": "SignedAssertionFilePath"
> }
> ]
2. Make a call to the WebApi running in AKS
3. Should see the attached error

### Error message

info: Microsoft.IdentityModel.LoggingExtensions.IdentityLoggerAdapter[0]
      IDX10242: Security token: '[PII of type 'System.String' is hidden. For more details, see [https://aka.ms/IdentityModel/PII.]'](https://aka.ms/IdentityModel/PII.]%27) has a valid signature.
info: Microsoft.IdentityModel.LoggingExtensions.IdentityLoggerAdapter[0]
      IDX10239: Lifetime of the token is valid.
info: Microsoft.IdentityModel.LoggingExtensions.IdentityLoggerAdapter[0]
      IDX10234: Audience Validated.Audience: 'd4bff1b1-ae02-4475-bdee-b08ac7daf110'
info: Microsoft.IdentityModel.LoggingExtensions.IdentityLoggerAdapter[0]
      IDX10245: Creating claims identity from the validated token: '[PII of type 'System.IdentityModel.Tokens.Jwt.JwtSecurityToken' is hidden. For more details, see [https://aka.ms/IdentityModel/PII.]'.](https://aka.ms/IdentityModel/PII.]%27.)
info: Microsoft.IdentityModel.LoggingExtensions.IdentityLoggerAdapter[0]
      IDX10241: Security token validated. token: '[PII of type 'System.String' is hidden. For more details, see [https://aka.ms/IdentityModel/PII.]'.](https://aka.ms/IdentityModel/PII.]%27.)
info: Microsoft.Identity.Web.TokenAcquisition[402]
      [MsIdWeb] Using Pod identity file not found as client credentials. 
info: Microsoft.Identity.Web.TokenAcquisition[300]
      [MsIdWeb] An error occured during token acquisition: IDW10501: Exception acquiring token for a confidential client. 
      System.ArgumentException: Delegate to an instance method cannot have null 'this'.
         at System.MulticastDelegate.ThrowNullThisInDelegateToInstance()
         at Microsoft.Identity.Web.ConfidentialClientApplicationBuilderExtension.WithClientCredentials(ConfidentialClientApplicationBuilder builder, IEnumerable`1 clientCredentials, ILogger logger, ICredentialsLoader credentialsLoader, CredentialSourceLoaderParameters credentialSourceLoaderParameters)
         at Microsoft.Identity.Web.TokenAcquisition.BuildConfidentialClientApplication(MergedOptions mergedOptions)
fail: Microsoft.AspNetCore.Server.Kestrel[13]
      Connection id "0HMQN5SQEMHBD", Request id "0HMQN5SQEMHBD:00000002": An unhandled exception was thrown by the application.
      Status Code: 0
      Microsoft.Graph.ServiceException: Code: generalException
      Message: An error occurred sending the request.

       ---> System.ArgumentException: Delegate to an instance method cannot have null 'this'.
         at System.MulticastDelegate.ThrowNullThisInDelegateToInstance()
         at Microsoft.Identity.Web.ConfidentialClientApplicationBuilderExtension.WithClientCredentials(ConfidentialClientApplicationBuilder builder, IEnumerable`1 clientCredentials, ILogger logger, ICredentialsLoader credentialsLoader, CredentialSourceLoaderParameters credentialSourceLoaderParameters)
         at Microsoft.Identity.Web.TokenAcquisition.BuildConfidentialClientApplication(MergedOptions mergedOptions)
         at Microsoft.Identity.Web.TokenAcquisition.GetOrBuildConfidentialClientApplication(MergedOptions mergedOptions)
         at Microsoft.Identity.Web.TokenAcquisition.GetAuthenticationResultForUserAsync(IEnumerable`1 scopes, String authenticationScheme, String tenantId, String userFlow, ClaimsPrincipal user, TokenAcquisitionOptions tokenAcquisitionOptions)
         at Microsoft.Identity.Web.DefaultAuthorizationHeaderProvider.CreateAuthorizationHeaderForUserAsync(IEnumerable`1 scopes, AuthorizationHeaderProviderOptions downstreamApiOptions, ClaimsPrincipal claimsPrincipal, CancellationToken cancellationToken)
         at Microsoft.Identity.Web.TokenAcquisitionAuthenticationProvider.AuthenticateRequestAsync(HttpRequestMessage request)
         at Microsoft.Graph.AuthenticationHandler.SendAsync(HttpRequestMessage httpRequestMessage, CancellationToken cancellationToken)
         at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
         at Microsoft.Graph.HttpProvider.SendRequestAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
         --- End of inner exception stack trace ---
         at Microsoft.Graph.HttpProvider.SendRequestAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
         at Microsoft.Graph.HttpProvider.SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
         at Microsoft.Graph.BaseRequest.SendRequestAsync(Object serializableObject, CancellationToken cancellationToken, HttpCompletionOption completionOption)
         at Microsoft.Graph.BaseRequest.SendAsync[T](Object serializableObject, CancellationToken cancellationToken, HttpCompletionOption completionOption)
         at Microsoft.Graph.UserRequest.GetAsync(CancellationToken cancellationToken)
         at AksTestWebApi.Controllers.WeatherForecastController.Get() in /src/AksTestWebApi/Controllers/WeatherForecastController.cs:line 31
         at lambda_method27(Closure , Object )
         at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.AwaitableObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
      --- End of stack trace from previous location ---
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|20_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
         at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
         at Microsoft.AspNetCore.Authorization.Policy.AuthorizationMiddlewareResultHandler.HandleAsync(RequestDelegate next, HttpContext context, AuthorizationPolicy policy, PolicyAuthorizationResult authorizeResult)
         at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
         at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)

### Id Web logs

_No response_

### Relevant code snippets

```csharp
NA
```

### Regression

_No response_

### Expected behavior

IdWeb should pick the federated token file from AKS environment and exchange it for a AAD access token.

### PR DESCRIPTION
**Pull request by jmprieur**
and the same issue for WIF for MSI
and adds loggin in WIF for AKS

### CODE REVIEW

File: src/Microsoft.Identity.Web.Certificate/DefaultCertificateLoader.cs [hunk 1]
```diff
@@ -26,6 +27,22 @@ namespace Microsoft.Identity.Web
...
     public class DefaultCertificateLoader : DefaultCredentialsLoader, ICertificateLoader
     {
+        /// <summary>
+        /// Constructor with a logger.
+        /// </summary>
+        /// <param name="logger"></param>
+        public DefaultCertificateLoader(ILogger<DefaultCertificateLoader>? logger) : base(logger)
+        {
+
+        }
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.Certificate/DefaultCertificateLoader.cs:37]: nit: remove line

File: src/Microsoft.Identity.Web.Certificate/DefaultCertificateLoader.cs [hunk 2]
```diff
@@ -26,6 +27,22 @@ namespace Microsoft.Identity.Web
...
+
+        /// <summary>
+        /// Default constuctor.
+        /// </summary>
+        //[Obsolete("Rather use the constructor with a logger")]
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.Certificate/DefaultCertificateLoader.cs:42]: >//[Obsolete("Rather use the constructor with a logger")] [](http://example.com/codeflow?start=8&length=56)

remove?

File: src/Microsoft.Identity.Web.Certificate/DefaultCredentialsLoader.cs [hunk 3]
```diff
@@ -12,20 +14,40 @@ namespace Microsoft.Identity.Web
...
+                { CredentialSource.Base64Encoded, new Base64EncodedCertificateLoader() },
+                { CredentialSource.SignedAssertionFromManagedIdentity, new SignedAssertionFromManagedIdentityCredentialLoader() },
+                { CredentialSource.SignedAssertionFilePath, new SignedAssertionFilePathCredentialsLoader(_logger) }
+            };
+        }
+
+        /// <summary>
+        /// Default constructor (for backward compatibility)
+        /// </summary>
+        //[Obsolete("Rather use the constructor with a logger.")]
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.Certificate/DefaultCredentialsLoader.cs:41]: >  //[Obsolete("Rather use the constructor with a logger.")] [](http://example.com/codeflow?start=6&length=59)

remove

File: src/Microsoft.Identity.Web.Certificateless/AzureIdentityForKubernetesClientAssertion.cs [hunk 4]
```diff
@@ -14,40 +15,79 @@ namespace Microsoft.Identity.Web
...
+        public AzureIdentityForKubernetesClientAssertion(string? filePath, ILogger? logger = null)
         {
+            _logger = logger;
+
+            if (filePath == null)
+            {
+                Log.SignedAssertionFileDiskPathNotProvided(_logger);
+            }
+
+            _filePath = _filePath ?? Environment.GetEnvironmentVariable("AZURE_ACCESS_TOKEN_FILE");
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.Certificateless/AzureIdentityForKubernetesClientAssertion.cs:44]: >AZURE_ACCESS_TOKEN_FILE [](http://example.com/codeflow?start=73&length=23)

can these be constants in this class? or a constants file?

File: src/Microsoft.Identity.Web.Certificateless/AzureIdentityForKubernetesClientAssertion.cs [hunk 5]
```diff
@@ -14,40 +15,79 @@ namespace Microsoft.Identity.Web
...
+            {
+                Log.SignedAssertionEnvironmentVariableNotProvided(_logger, "AZURE_FEDERATED_TOKEN_FILE");
+                Log.NoSignedAssertionParameterProvided(_logger);
+            }
         }
 
-        private readonly string _filePath;
+        private readonly string? _filePath;
+
+        private readonly ILogger? _logger;
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.Certificateless/AzureIdentityForKubernetesClientAssertion.cs:61]: move to top of the class

File: src/Microsoft.Identity.Web.Certificateless/AzureIdentityForKubernetesClientAssertion.Log.cs [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+            [LoggerMessage(EventId = 6, Level = LogLevel.Error, Message = "The file `{filePath} does not contain a valid signed assertion. {message}.")]
+            partial void FileDoesNotContainValidAssertion(ILogger logger, string filePath, string message);
+    #pragma warning restore SYSLIB1009 // Logging methods must be static
+    */
+
+        /// <summary>
+        /// Performant logging messages.
+        /// </summary>
+        static class Log
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.Certificateless/AzureIdentityForKubernetesClientAssertion.Log.cs:37]: rename to Logger, as well as the class, to match our other logger classes.

File: src/Microsoft.Identity.Web.Certificateless/AzureIdentityForKubernetesClientAssertion.Log.cs [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        {
+            private static readonly Action<ILogger, Exception?> __SignedAssertionFileDiskPathNotProvidedCallback =
+                LoggerMessage.Define(LogLevel.Information, new EventId(1, nameof(SignedAssertionFileDiskPathNotProvided)), "SignedAssertionFileDiskPath not provided. Falling back to the content of the AZURE_FEDERATED_TOKEN_FILE environment variable.");
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.Certificateless/AzureIdentityForKubernetesClientAssertion.Log.cs:40]: >. [](http://example.com/codeflow?start=248&length=1)

nit: add a space after all the periods.

File: src/Microsoft.Identity.Web.Certificateless/AzureIdentityForKubernetesClientAssertion.Log.cs [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+                LoggerMessage.Define<string, string>(LogLevel.Error, new EventId(6, nameof(FileDoesNotContainValidAssertion)), "The file `{filePath} does not contain a valid signed assertion. {message}.");
+
+            public static void FileDoesNotContainValidAssertion(ILogger? logger, string filePath, string message)
+            {
+                if (logger != null && logger.IsEnabled(LogLevel.Error))
+                {
+                    __FileDoesNotContainValidAssertionCallback(logger, filePath, message, null);
+                }
+            }
+        }
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.Certificateless/AzureIdentityForKubernetesClientAssertion.Log.cs:99]: can you separate the private methods from the public ones. Here's an example: [https://github.com/AzureAD/microsoft-identity-web/blob/master/src/Microsoft.Identity.Web.TokenCache/Distributed/MsalDistributedTokenCacheAdapter.Logger.cs#L114](https://github.com/AzureAD/microsoft-identity-web/blob/master/src/Microsoft.Identity.Web.TokenCache/Distributed/MsalDistributedTokenCacheAdapter.Logger.cs#L114) 

and the public ones will need xml comments.

File: tests/Microsoft.Identity.Web.Test/AzureIdentityForKubernetesClientAssertionTests.cs [hunk 9]
```diff
@@ -0,0 +1,69 @@
+using System;
```
Comment (jennyf19) [inline tests/Microsoft.Identity.Web.Test/AzureIdentityForKubernetesClientAssertionTests.cs:1]: license header info is missing...it should be added by default when you create a class now :(

File: tests/Microsoft.Identity.Web.Test/SignedAssertionFilePathCredentialsLoader.cs [hunk 10]
```diff
@@ -0,0 +1,81 @@
+using System;
```
Comment (jennyf19) [inline tests/Microsoft.Identity.Web.Test/SignedAssertionFilePathCredentialsLoader.cs:1]: license info missing.

File: tests/Microsoft.Identity.Web.Test/SignedAssertionFilePathCredentialsLoader.cs [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+using System.Threading.Tasks;
+using Microsoft.Identity.Abstractions;
+using Microsoft.IdentityModel.JsonWebTokens;
+using Xunit;
+
+namespace Microsoft.Identity.Web.Tests.Certificateless
+{
+    public class SignedAssertionFilePathCredentialsLoaderTests
+    {
+        const string filePath = "signedAssertion.txt";
```
Comment (jennyf19) [inline tests/Microsoft.Identity.Web.Test/SignedAssertionFilePathCredentialsLoader.cs:13]: >"signedAssertion.txt" [](http://example.com/codeflow?start=32&length=21)

move to the TestConstants.cs file

File: tests/Microsoft.Identity.Web.Test/SignedAssertionFilePathCredentialsLoader.cs [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+        string token;
+        SignedAssertionFilePathCredentialsLoader signedAssertionFilePathCredentialsLoader = new SignedAssertionFilePathCredentialsLoader(null);
+
+
```
Comment (jennyf19) [inline tests/Microsoft.Identity.Web.Test/SignedAssertionFilePathCredentialsLoader.cs:17]: remove line

**Review by jennyf19 — CHANGES_REQUESTED** [review]
:clock1:

File: src/Microsoft.Identity.Web.Certificate/DefaultCertificateLoader.cs (hunk 1 above)
Comment (jmprieur) [reply src/Microsoft.Identity.Web.Certificate/DefaultCertificateLoader.cs:37]: Thanks @jennyf19 . Done!

File: src/Microsoft.Identity.Web.Certificate/DefaultCertificateLoader.cs (hunk 2 above)
Comment (jmprieur) [reply src/Microsoft.Identity.Web.Certificate/DefaultCertificateLoader.cs:42]: Done

File: src/Microsoft.Identity.Web.Certificate/DefaultCredentialsLoader.cs (hunk 3 above)
Comment (jmprieur) [reply src/Microsoft.Identity.Web.Certificate/DefaultCredentialsLoader.cs:41]: Done

File: src/Microsoft.Identity.Web.Certificateless/AzureIdentityForKubernetesClientAssertion.Log.cs (hunk 6 above)
Comment (jmprieur) [reply src/Microsoft.Identity.Web.Certificateless/AzureIdentityForKubernetesClientAssertion.Log.cs:37]: Done!

File: src/Microsoft.Identity.Web.Certificateless/AzureIdentityForKubernetesClientAssertion.Log.cs (hunk 7 above)
Comment (jmprieur) [reply src/Microsoft.Identity.Web.Certificateless/AzureIdentityForKubernetesClientAssertion.Log.cs:40]: Done

File: src/Microsoft.Identity.Web.Certificateless/AzureIdentityForKubernetesClientAssertion.Log.cs (hunk 8 above)
Comment (jmprieur) [reply src/Microsoft.Identity.Web.Certificateless/AzureIdentityForKubernetesClientAssertion.Log.cs:99]: Logger is a private class?

File: src/Microsoft.Identity.Web.Certificateless/AzureIdentityForKubernetesClientAssertion.cs (hunk 4 above)
Comment (jmprieur) [reply src/Microsoft.Identity.Web.Certificateless/AzureIdentityForKubernetesClientAssertion.cs:44]: Done!

File: tests/Microsoft.Identity.Web.Test/AzureIdentityForKubernetesClientAssertionTests.cs (hunk 9 above)
Comment (jmprieur) [reply tests/Microsoft.Identity.Web.Test/AzureIdentityForKubernetesClientAssertionTests.cs:1]: fixed

File: tests/Microsoft.Identity.Web.Test/SignedAssertionFilePathCredentialsLoader.cs (hunk 11 above)
Comment (jmprieur) [reply tests/Microsoft.Identity.Web.Test/SignedAssertionFilePathCredentialsLoader.cs:13]: done

**Review by jennyf19 — APPROVED** [review]
:shipit:

**Pull request merged**